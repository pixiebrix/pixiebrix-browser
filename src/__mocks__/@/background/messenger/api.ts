/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// This mock file provides reasonable defaults for the methods in the background messenger API

/* Do not use `registerMethod` in this file */
import {
  backgroundTarget as bg,
  getMethod,
  getNotifier,
} from "webext-messenger";
import type { AxiosRequestConfig } from "axios";
import type { RemoteResponse } from "@/types/contract";
import { uuidv4 } from "@/types/helpers";
import { SanitizedIntegrationConfig } from "@/types/integrationTypes";
import { RegistryId } from "@/types/registryTypes";

// Chrome offers this API in more contexts than Firefox, so it skips the messenger entirely
export const containsPermissions = jest
  .fn()
  .mockRejectedValue(
    new Error(
      "Internal mocking error: jest should be using containsPermissions from mockPermissions"
    )
  );

export const getAvailableVersion = getMethod("GET_AVAILABLE_VERSION", bg);
export const ensureContentScript = getMethod("INJECT_SCRIPT", bg);
export const openPopupPrompt = getMethod("OPEN_POPUP_PROMPT", bg);
export const getUID = jest.fn().mockResolvedValue(uuidv4());

export const pong = jest.fn(() => ({
  timestamp: Date.now(),
}));

export const waitForTargetByUrl = getMethod("WAIT_FOR_TARGET_BY_URL", bg);

export const activatePartnerTheme = getMethod("ACTIVATE_PARTNER_THEME", bg);
export const getPartnerPrincipals = getMethod("GET_PARTNER_PRINCIPALS", bg);
export const launchAuthIntegration = getMethod("LAUNCH_AUTH_INTEGRATION", bg);

export const activateTab = getMethod("ACTIVATE_TAB", bg);
export const reactivateEveryTab = getNotifier("REACTIVATE_EVERY_TAB", bg);

export const closeTab = getMethod("CLOSE_TAB", bg);
export const deleteCachedAuthData = getMethod("DELETE_CACHED_AUTH", bg);
export const getCachedAuthData = getMethod("GET_CACHED_AUTH", bg);
export const clearServiceCache = getMethod("CLEAR_SERVICE_CACHE", bg);
export const sheets = {
  getAllSpreadsheets: jest.fn().mockRejectedValue(new Error("Not implemented")),
  getSpreadsheet: jest.fn().mockRejectedValue(new Error("Not implemented")),
  getTabNames: jest.fn().mockRejectedValue(new Error("Not implemented")),
  getSheetProperties: jest.fn().mockRejectedValue(new Error("Not implemented")),
  getHeaders: jest.fn().mockRejectedValue(new Error("Not implemented")),
  getAllRows: jest.fn().mockRejectedValue(new Error("Not implemented")),
  createTab: getMethod("GOOGLE_SHEETS_CREATE_TAB", bg),
  appendRows: getMethod("GOOGLE_SHEETS_APPEND_ROWS", bg),
};

/**
 * Uninstall context menu and return whether the context menu was uninstalled.
 */
export const uninstallContextMenu = getMethod("UNINSTALL_CONTEXT_MENU", bg);
export const ensureContextMenu = getMethod("ENSURE_CONTEXT_MENU", bg);
export const openTab = getMethod("OPEN_TAB", bg);

export const registry = {
  fetch: jest.fn().mockResolvedValue(true),
  syncRemote: getMethod("REGISTRY_SYNC", bg),
  getByKinds: jest.fn().mockResolvedValue([]),
  find: jest.fn().mockImplementation(async (id: RegistryId) => {
    throw new Error(
      `Find not implemented in registry mock (looking up "${id}"). See __mocks__/background/messenger/api for more information.`
    );
  }),
  clear: getMethod("REGISTRY_CLEAR", bg),
};

export const dataStore = {
  get: jest.fn().mockRejectedValue(new Error("Not implemented in mock")),
  set: getMethod("SET_DATA_STORE", bg),
};

export const requestRun = {
  inOpener: getMethod("REQUEST_RUN_IN_OPENER", bg),
  inTarget: getMethod("REQUEST_RUN_IN_TARGET", bg),
  inTop: getMethod("REQUEST_RUN_IN_TOP", bg),
  inOtherTabs: getMethod("REQUEST_RUN_IN_OTHER_TABS", bg),
};

export const contextMenus = {
  preload: getMethod("PRELOAD_CONTEXT_MENUS", bg),
};

export const services = {
  locateAllForId: jest.fn().mockResolvedValue([]),
  locate: jest
    .fn()
    .mockRejectedValue(new Error("Locate not implemented in mock")),
  refresh: getMethod("REFRESH_SERVICES", bg),
  refreshLocal: getMethod("LOCATOR_REFRESH_LOCAL", bg),
};

// `getMethod` currently strips generics, so we must copy the function signature here
export const proxyService = getMethod("PROXY", bg) as <TData>(
  serviceConfig: SanitizedIntegrationConfig | null,
  requestConfig: AxiosRequestConfig
) => Promise<RemoteResponse<TData>>;

// Use this instead: `import reportError from "@/telemetry/reportError"`
// export const recordError = getNotifier("RECORD_ERROR", bg);

export const recordLog = getNotifier("RECORD_LOG", bg);
export const recordWarning = getNotifier("RECORD_WARNING", bg);
export const recordEvent = getNotifier("RECORD_EVENT", bg);
export const recordBrickRun = getNotifier("RECORD_BRICK_RUN", bg);
export const clearLogs = getMethod("CLEAR_LOGS", bg);
export const clearLog = getMethod("CLEAR_LOG", bg);
export const clearExtensionDebugLogs = getMethod(
  "CLEAR_EXTENSION_DEBUG_LOGS",
  bg
);

export const traces = {
  addEntry: getNotifier("ADD_TRACE_ENTRY", bg),
  addExit: getNotifier("ADD_TRACE_EXIT", bg),
  clear: getMethod("CLEAR_TRACES", bg),
  clearAll: getNotifier("CLEAR_ALL_TRACES", bg),
};

export const initTelemetry = getNotifier("INIT_TELEMETRY", bg);
export const sendDeploymentAlert = getNotifier("SEND_DEPLOYMENT_ALERT", bg);

export const captureTab = getMethod("CAPTURE_TAB", bg);

export const getUserData = getMethod("GET_USER_DATA", bg);
