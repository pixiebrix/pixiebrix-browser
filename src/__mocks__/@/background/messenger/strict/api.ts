/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { SanitizedIntegrationConfig } from "@/integrations/integrationTypes";
import { RemoteResponse } from "@/types/contract";
import { NetworkRequestConfig } from "@/types/networkTypes";
import { RegistryId } from "@/types/registryTypes";
import { Nullishable } from "@/utils/nullishUtils";
import { getMethod, backgroundTarget as bg } from "webext-messenger";

export * from "../../../../../background/messenger/strict/api";

export const registry = {
  syncRemote: jest.fn(),
  getByKinds: jest.fn().mockResolvedValue([]),
  find: jest.fn().mockImplementation(async (id: RegistryId) => {
    throw new Error(
      `Find not implemented in registry mock (looking up "${id}"). See __mocks__/background/messenger/api for more information.`,
    );
  }),
  clear: getMethod("REGISTRY_CLEAR", bg),
};

export const services = {
  locateAllForId: jest.fn().mockResolvedValue([]),
  locate: jest
    .fn()
    .mockRejectedValue(new Error("Locate not implemented in mock")),
  refresh: jest.fn(),
  refreshLocal: getMethod("LOCATOR_REFRESH_LOCAL", bg),
};

// `getMethod` currently strips generics, so we must copy the function signature here
export const performConfiguredRequestInBackground = getMethod(
  "CONFIGURED_REQUEST",
  bg,
) as <TData>(
  integrationConfig: Nullishable<SanitizedIntegrationConfig>,
  requestConfig: NetworkRequestConfig,
) => Promise<RemoteResponse<TData>>;

export const dataStore = {
  get: jest.fn().mockRejectedValue(new Error("Not implemented in mock")),
  set: getMethod("SET_DATA_STORE", bg),
};
