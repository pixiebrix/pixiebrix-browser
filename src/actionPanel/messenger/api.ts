/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/* Do not use `registerMethod` in this file */
import { getMethod } from "webext-messenger";
import { isBrowserActionPanel } from "@/chrome";

// TODO: This should be a hard error, but due to unknown dependency routes, it can't be enforced yet
if (isBrowserActionPanel() && process.env.DEBUG) {
  console.warn(
    "This should not have been imported in the action panel. Use the API directly instead."
  );
}

export const renderPanels = getMethod("ACTION_PANEL_RENDER_PANELS");
export const showForm = getMethod("ACTION_PANEL_SHOW_FORM");
export const hideForm = getMethod("ACTION_PANEL_HIDE_FORM");
