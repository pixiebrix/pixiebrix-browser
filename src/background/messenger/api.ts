/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/* Do not use `registerMethod` in this file */
import { getMethod } from "webext-messenger";
import { browser } from "webextension-polyfill-ts";
import { isBackgroundPage } from "webext-detect-page";

// TODO: This should be a hard error, but due to unknown dependency routes, it can't be enforced yet
if (isBackgroundPage()) {
  console.trace(
    "This should not have been imported in the background page. Use the API directly instead."
  );
}

// Chrome offers this API in more contexts than Firefox, so it skips the messenger entirely
export const containsPermissions = browser.permissions
  ? browser.permissions.contains
  : getMethod("CONTAINS_PERMISSIONS");

export const openPopupPrompt = getMethod("OPEN_POPUP_PROMPT");
export const whoAmI = getMethod("ECHO_SENDER");
export const activateTab = getMethod("ACTIVATE_TAB");
export const closeTab = getMethod("CLOSE_TAB");
export const markTabAsReady = getMethod("MARK_TAB_AS_READY");

/**
 * Uninstall context menu and return whether or not the context menu was uninstalled.
 */
export const uninstallContextMenu = getMethod("UNINSTALL_CONTEXT_MENU");
export const ensureContextMenu = getMethod("ENSURE_CONTEXT_MENU");
export const openTab = getMethod("OPEN_TAB");

// Temporary, webext-messenger depends on this global
(globalThis as any).browser = browser;
