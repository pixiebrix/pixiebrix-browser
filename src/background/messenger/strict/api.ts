/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/* Do not use `registerMethod` in this file */
import {
  backgroundTarget as bg,
  getMethod,
  getNotifier,
} from "webext-messenger";

export const showMySidePanel = getMethod("SHOW_MY_SIDE_PANEL", bg);
export const ensureContentScript = getMethod("INJECT_SCRIPT", bg);

export const dataStore = {
  get: getMethod("GET_DATA_STORE", bg),
  set: getMethod("SET_DATA_STORE", bg),
};
export const activatePartnerTheme = getMethod("ACTIVATE_PARTNER_THEME", bg);

export const traces = {
  addEntry: getNotifier("ADD_TRACE_ENTRY", bg),
  addExit: getNotifier("ADD_TRACE_EXIT", bg),
  clear: getMethod("CLEAR_TRACES", bg),
  clearAll: getNotifier("CLEAR_ALL_TRACES", bg),
};

export const captureTab = getMethod("CAPTURE_TAB", bg);
export const deleteCachedAuthData = getMethod("DELETE_CACHED_AUTH", bg);
export const getCachedAuthData = getMethod("GET_CACHED_AUTH", bg);
export const setToolbarBadge = getMethod("SET_TOOLBAR_BADGE", bg);
