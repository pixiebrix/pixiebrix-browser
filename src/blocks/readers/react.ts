/*
 * Copyright (C) 2020 Pixie Brix, LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import { createSendScriptMessage } from "@/messaging/chrome";
import { READ_REACT_COMPONENT } from "@/messaging/constants";
import { ReaderOutput } from "@/core";
import { registerFactory } from "@/blocks/readers/factory";

export interface ReactConfig {
  type: "react";
  selector: string;
  traverseUp?: number;
  rootProp?: string;
  waitMillis?: number;
}

export const withReactComponent = createSendScriptMessage<ReaderOutput>(
  READ_REACT_COMPONENT
);

async function doRead(reader: ReactConfig): Promise<ReaderOutput> {
  const { selector, traverseUp = 0, waitMillis = 1000, rootProp } = reader;
  return await withReactComponent({
    selector,
    traverseUp,
    waitMillis,
    rootProp,
  });
}

registerFactory("react", doRead);
