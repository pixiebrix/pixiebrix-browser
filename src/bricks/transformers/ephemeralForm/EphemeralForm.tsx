/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import {
  cancelForm,
  getFormDefinition,
} from "@/contentScript/messenger/strict/api";
import Loader from "@/components/Loader";
import { getErrorMessage } from "@/errors/errorHelpers";
import { type Target } from "@/types/messengerTypes";
import { validateUUID } from "@/types/helpers";
import { TOP_LEVEL_FRAME_ID } from "@/domConstants";
import useAsyncState from "@/hooks/useAsyncState";
import ErrorBoundary from "@/components/ErrorBoundary";
import useReportError from "@/hooks/useReportError";
import IsolatedComponent from "@/components/IsolatedComponent";
import { type EphemeralFormContentProps } from "./EphemeralFormContent";
import { assertNotNullish } from "@/utils/nullishUtils";

const ModalLayout: React.FC = ({ children }) => (
  // Don't use React Bootstrap's Modal because we want to customize the classes in the layout
  <div className="modal-content">
    <div className="modal-body">{children}</div>
  </div>
);

const PanelLayout: React.FC = ({ children }) => (
  <div className="p-3">{children}</div>
);

const IsolatedEphemeralFormContent: React.FunctionComponent<
  EphemeralFormContentProps
> = (props) => (
  <IsolatedComponent
    name="EphemeralFormContent"
    noStyle={props.definition.disableParentStyles}
    lazy={async () =>
      import(
        /* webpackChunkName: "isolated/EphemeralFormContent" */
        "./EphemeralFormContent"
      )
    }
    factory={(EphemeralFormContent) => <EphemeralFormContent {...props} />}
  />
);

function validateOpener(opener: string | null): Target {
  if (opener == null) {
    throw new Error("Missing opener");
  }

  try {
    const parsed = JSON.parse(opener);
    if (
      parsed &&
      typeof parsed === "object" &&
      "tabId" in parsed &&
      "frameId" in parsed
    ) {
      return parsed as Target;
    }

    throw new TypeError(`Invalid opener: ${opener}`);
  } catch {
    throw new TypeError(`Invalid opener: ${opener}`);
  }
}

/**
 * @see FormTransformer
 */
const EphemeralForm: React.FC = () => {
  const params = new URLSearchParams(location.search);
  const nonce = validateUUID(params.get("nonce"));
  const opener = validateOpener(params.get("opener"));
  const mode = params.get("mode") ?? "modal";

  const isModal = mode === "modal";

  // The opener for a sidebar panel will be the sidebar frame, not the host panel frame. The sidebar only opens in the
  // top-level frame, so hard-code the top-level frameId
  const target = isModal
    ? opener
    : { tabId: opener.tabId, frameId: TOP_LEVEL_FRAME_ID };
  const FormContainer = isModal ? ModalLayout : PanelLayout;

  const {
    data: formDefinition,
    isLoading,
    error,
  } = useAsyncState(async () => getFormDefinition(target, nonce), [nonce]);

  useReportError(error);

  if (isLoading) {
    return (
      <FormContainer>
        <Loader />
      </FormContainer>
    );
  }

  if (error) {
    return (
      <FormContainer>
        <div>Form Error</div>

        <div className="text-danger my-3">{getErrorMessage(error)}</div>

        <div>
          <button
            className="btn btn-primary"
            type="button"
            onClick={() => {
              cancelForm(target, nonce);
            }}
          >
            Close
          </button>
        </div>
      </FormContainer>
    );
  }

  assertNotNullish(formDefinition, "unable to load form definition");

  return (
    <FormContainer>
      <ErrorBoundary>
        <IsolatedEphemeralFormContent
          definition={formDefinition}
          target={target}
          nonce={nonce}
          isModal={isModal}
        />
      </ErrorBoundary>
    </FormContainer>
  );
};

export default EphemeralForm;
