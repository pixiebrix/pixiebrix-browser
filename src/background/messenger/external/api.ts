/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * @file API for the PixieBrix app to talk to the browser extension.
 * TODO: It's not yet using the Messenger: https://github.com/pixiebrix/webext-messenger/issues/6
 */

import { _liftBackground } from "@/background/externalProtocol";
import * as local from "@/background/messenger/external/_implementation";
// eslint-disable-next-line import/no-restricted-paths -- Legacy code, needs https://github.com/pixiebrix/webext-messenger/issues/6
import { liftExternalToContentScript } from "@/contentScript/externalProtocol";
import { isChrome } from "webext-detect-page";
import { type SerializableResponse } from "@/pageScript/messenger/pigeon";
import { readPartnerAuthData } from "@/auth/token";

const liftExternal = isChrome()
  ? // Chrome can communicate directly via the standard chrome.runtime.sendMessage API.
    _liftBackground
  : // Firefox doesn't support web-to-background communication, so it must be travel via the content script.
    <TArguments extends unknown[], R extends SerializableResponse>(
      type: string,
      method: (...args: TArguments) => Promise<R>
    ) => {
      const liftedToBackground = _liftBackground(`BACKGROUND_${type}`, method);
      return liftExternalToContentScript(type, liftedToBackground);
    };

export const connectPage = liftExternal("CONNECT_PAGE", async () =>
  browser.runtime.getManifest()
);

export const setExtensionAuth = liftExternal(
  "SET_EXTENSION_AUTH",
  local.setExtensionAuth
);

export const setActivatingBlueprint = liftExternal(
  "SET_ACTIVATING_BLUEPRINT",
  local.setActivatingBlueprint
);

export const openMarketplace = liftExternal(
  "OPEN_MARKETPLACE",
  local.openMarketplace
);

export const openActivateBlueprint = liftExternal(
  "OPEN_ACTIVATE_BLUEPRINT",
  local.openActivateBlueprint
);

export const openExtensionOptions = liftExternal(
  "OPEN_OPTIONS",
  local.openExtensionOptions
);

export const getPartnerToken = liftExternal(
  "GET_PARTNER_TOKEN",
  readPartnerAuthData
);
