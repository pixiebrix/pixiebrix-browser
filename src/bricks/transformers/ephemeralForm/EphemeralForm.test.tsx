/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { getFormDefinition } from "@/contentScript/messenger/strict/api";
import { render, screen } from "@testing-library/react";
import React from "react";
import EphemeralForm from "./EphemeralForm";

jest.mock("@/contentScript/messenger/strict/api");

const getFormDefinitionMock = jest.mocked(getFormDefinition);

describe("EphemeralForm", () => {
  it("shows field titles", async () => {
    getFormDefinitionMock.mockResolvedValue({
      schema: {
        title: "Test Form",
        type: "object",
        properties: {
          foo: {
            type: "string",
            title: "Foo",
          },
        },
      },
      uiSchema: {},
      cancelable: false,
      submitCaption: "Submit",
      location: "modal",
    });

    render(<EphemeralForm />);

    await expect(
      screen.findByRole("textbox", { name: /foo/i }),
    ).resolves.toBeVisible();
  });

  it("supports markdown in field descriptions", async () => {
    getFormDefinitionMock.mockResolvedValue({
      schema: {
        title: "Test Form",
        type: "object",
        properties: {
          foo: {
            title: "Foo",
            description: "I am **bold**",
            type: "string",
          },
        },
      },
      uiSchema: {},
      cancelable: false,
      submitCaption: "Submit",
      location: "modal",
    });

    render(<EphemeralForm />);

    await expect(
      screen.findByRole("heading", { level: 5, name: /test form/i }),
    ).resolves.toBeInTheDocument();
    expect(
      screen.getByText((_, element) => element.textContent === "I am bold"),
    ).toBeInTheDocument();
    expect(screen.getByRole("strong")).toHaveTextContent("bold");
  });

  it("supports markdown in form description", async () => {
    getFormDefinitionMock.mockResolvedValue({
      schema: {
        title: "Test Form",
        description: "I am **bold**",
        type: "object",
        properties: {},
      },
      uiSchema: {},
      cancelable: false,
      submitCaption: "Submit",
      location: "modal",
    });

    render(<EphemeralForm />);

    await expect(
      screen.findByRole("heading", { level: 5, name: /test form/i }),
    ).resolves.toBeInTheDocument();
    expect(
      screen.getByText((_, element) => element.textContent === "I am bold"),
    ).toBeInTheDocument();
    expect(screen.getByRole("strong")).toHaveTextContent("bold");
  });

  it("renders a text input with inputmode numeric in place of a number input", async () => {
    getFormDefinitionMock.mockResolvedValue({
      schema: {
        title: "Test Form",
        type: "object",
        properties: {
          rating: { type: "number", title: "Rating" },
        },
      },
      uiSchema: {},
      cancelable: false,
      submitCaption: "Submit",
      location: "modal",
    });

    render(<EphemeralForm />);

    await expect(
      screen.findByRole("textbox", { name: "Rating", hidden: true }),
    ).resolves.toHaveAttribute("inputmode", "numeric");

    expect(screen.queryByRole("spinButton")).not.toBeInTheDocument();
  });
});
