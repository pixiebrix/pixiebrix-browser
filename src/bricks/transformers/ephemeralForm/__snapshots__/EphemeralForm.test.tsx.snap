// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EphemeralForm shows field titles 1`] = `
<DocumentFragment>
  <div
    class="modal-content"
  >
    <div
      class="modal-body"
    >
      <form
        class="rjsf"
      >
        <div
          class="form-group"
        >
          <div
            class="my-1"
          >
            <h5>
              Test Form
            </h5>
            <hr
              class="border-0 bg-secondary"
              style="height: 1px;"
            />
          </div>
          <div
            class="p-0 container-fluid"
          >
            <div
              class="row"
              style="margin-bottom: 10px;"
            >
              <div
                class="col-12"
              >
                 
                <div
                  class="form-group"
                >
                  <div
                    class="mb-0 form-group"
                  >
                    <label
                      class="form-label"
                    >
                      foo
                    </label>
                    <input
                      class="form-control"
                      id="root_foo"
                      placeholder=""
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div>
          <button
            class="btn btn-primary"
            type="submit"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`EphemeralForm supports markdown in field descriptions 1`] = `
<DocumentFragment>
  <div
    class="modal-content"
  >
    <div
      class="modal-body"
    >
      <form
        class="rjsf"
      >
        <div
          class="form-group"
        >
          <div
            class="my-1"
          >
            <h5>
              Test Form
            </h5>
            <hr
              class="border-0 bg-secondary"
              style="height: 1px;"
            />
          </div>
          <div
            class="p-0 container-fluid"
          >
            <div
              class="row"
              style="margin-bottom: 10px;"
            >
              <div
                class="col-12"
              >
                 
                <div
                  class="form-group"
                >
                  <div
                    class="mb-0 form-group"
                  >
                    <label
                      class="form-label"
                    >
                      Foo
                    </label>
                    <input
                      class="form-control"
                      id="root_foo"
                      placeholder=""
                      type="text"
                      value=""
                    />
                  </div>
                  <div
                    class="field-description text-muted"
                    id="root_foo"
                  >
                    <div>
                      <p>
                        I am 
                        <strong>
                          bold
                        </strong>
                      </p>
                      

                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div>
          <button
            class="btn btn-primary"
            type="submit"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`EphemeralForm supports markdown in form description 1`] = `
<DocumentFragment>
  <div
    class="modal-content"
  >
    <div
      class="modal-body"
    >
      <form
        class="rjsf"
      >
        <div
          class="form-group"
        >
          <div
            class="my-1"
          >
            <h5>
              Test Form
            </h5>
            <hr
              class="border-0 bg-secondary"
              style="height: 1px;"
            />
          </div>
          <div
            class="field-description"
            id="root-description"
          >
            <div>
              <p>
                I am 
                <strong>
                  bold
                </strong>
              </p>
              

            </div>
          </div>
          <div
            class="p-0 container-fluid"
          />
        </div>
        <div>
          <button
            class="btn btn-primary"
            type="submit"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
